0021/README.md

```diff
@@ -0,0 +1,155 @@
+---
+name: Tenant Metadata Management Protocol
+shortname: 21/TMMP
+status: draft
+editor: Andrew Pasika <apasika@kaaiot.io>
+---
+
+<!-- toc -->
+
+
+# Introduction
+
+Tenant Metadata Management Protocol (TMMP) is designed to communicate tenant metadata between Kaa services.
+
+TMMP complies with the [Inter-Service Messaging](/0003/README.md) guidelines.
+
+
+# Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+The following terms and definitions are used in this RFC.
+
+- **Tenant metadata repository (repository)**: service that exposes TMMP to other services for managing tenant metadata.
+
+- **Tenant metadata client (client)**: service that uses TMMP interface exposed by a repository to manage tenant metadata.
+
+
+# Design
+
+## Tenant metadata get request
+
+*Tenant metadata get request* message is a [targeted message](/0003/README.md#targeted-messaging) that client sends to repository to retrieve the tenant metadata.
+
+The client MUST send tenant metadata get request messages using the following NATS subject:
+
+```
+kaa.v1.service.{repository-service-instance-name}.tmmp.tenant-metadata-get-request
+```
+
+The client MUST include NATS `replyTo` field to handle the response.
+It is RECOMMENDED to follow the subject format described in [3/ISM session affinity section](/0003/README.md#session-affinity):
+
+```
+kaa.v1.replica.{client-service-replica-id}.tmmp.tenant-metadata-get-response
+```
+
+Tenant metadata get request message payload MUST be an [Avro-encoded](https://avro.apache.org/) object with the following schema ([0021-tenant-metadata-get-request.avsc](./0021-tenant-metadata-get-request.avsc)):
+
+```json
+{
+    "namespace":"org.kaaproject.ipc.tmmp.gen.v1",
+    "name":"TenantMetadataGetRequest",
+    "type":"record",
+    "doc":"Tenant metadata get request",
+    "fields":[
+        {
+            "name":"correlationId",
+            "type":"string",
+            "doc":"Message ID primarily used to track message processing across services"
+        },
+        {
+            "name":"timestamp",
+            "type":"long",
+            "doc":"Message creation UNIX timestamp in milliseconds"
+        },
+        {
+            "name":"timeout",
+            "type":"long",
+            "default":0,
+            "doc":"Amount of milliseconds (since the timestamp) until the message expires. Value of 0 is reserved to indicate no expiration."
+        },
+        {
+            "name":"appName",
```

Comment:
Why query tenant metadata by app/appV? Should not services know from Tekton, which tenantID the app/appV belongs to?

Reply:
Currently, platform events contain appName / appVersion from which we should resolve tenant ID. The TMMP is intended to be implemented by Tekton.